vulnerabilities/data_source.py

```diff
@@ -380,3 +390,127 @@ def _include_file(
         match = match and path.endswith(f'.{file_ext}')
 
     return match
+
+
+class OvalDataSource(DataSource):
+    """
+    All data sources which collect data from OVAL files must inherit from this
+    `OvalDataSource` class. Subclasses must implement the methods `_fetch` and `set_api`.
+    """
+    @staticmethod
+    def create_purl(pkg_name: str, pkg_version: str, pkg_data: Mapping) -> PackageURL:
+        """
+        Helper method for creating different purls for subclasses without them reimplementing
+        get_data_from_xml_doc  method
+        Note: pkg_data must include 'type' of package
+        """
+        return PackageURL(name=pkg_name, version=pkg_version, **pkg_data)
+
+    @staticmethod
+    def _collect_pkgs(parsed_oval_data: Mapping) -> Set:
+        """
+        Helper method, used for loading the API. It expects data from
+        OvalParser.get_data().
+        """
+        all_pkgs = set()
+        for definition_data in parsed_oval_data:
+            for test_data in definition_data['test_data']:
+                for package in test_data['package_list']:
+                    all_pkgs.add(package)
+
+        return all_pkgs
+
+    def _fetch() -> Tuple[Mapping, Iterable[ET.ElementTree]]:
+        """
+        This method  contains logic to fetch OVAL files and yield them into
+        a tuple of file's metadata and it's ET.ElementTree.
+        Subclasses must implement this method.
+
+        Note: Mapping MUST INCLUDE "type" key. Example values of Mapping
+              {"type":"deb","qualifiers":{"distro":"buster"} }
+
+        """
+        raise NotImplementedError
+
+    def added_advisories(self) -> List[Advisory]:
```

Comment:
`added_advisories()` *must* only return advisories that have been added since the last run of the importer. All other advisories should be returned by `updated_advisories()`. Is this somehow ensured elsewhere in your implementation?

Reply:
Oops, I forgot to change it back to `updated_advisories`
No the implementation, doesn't do that. I will change this to `updated_advisories`, I used `added_advisories` for fast testing because it uses bulk inserts.

Reply:
I see. From a quick glance at the OVAL test data, it seems advisories have an attribute `public_date`. Maybe we can use that to tell added and updated advisories apart. But that can be worked on in a separate PR.

Reply:
No, they are for a  whole  xml oval document, which results into creation of 1000s of `Advisory` objects. No way to know which `Advisory` is changed.

Reply:
Oh, I see what you mean,  apparently Ubuntu OVAL files provide `public_date` , seems vendor specific since SUSE OVAL files don't contain this. 

This is good news, I agree with you let's first get this merged and I'll iterate upon it in other small PR.